### PR TITLE
Partition Table, Stages, Mounts, and Devices generator commands for otk

### PR DESCRIPTION
**Pull request by achilleas-k**
This is a joint PR by @mvo5 and me that adds three new commands:
- otk-gen-partition-table: generates a fully resolved partition table from a base description and modifications.  It's an externalised version of the `disk.NewPartitionTable()` function.
- otk-make-partition-stages: generates stages that can be added to a manifest to create a disk image from a fully resolved partition table.  It's an externalised version of the `osbuild.GenImagePrepareStages()` function.
- otk-make-partition-mounts-devices: generates mounts and devices that can be attached to osbuild stages that need to operate on a partitioned image file.  It's an externalised version of the `osbuild.GenMountsDevicesFromPT()` function.

Some tests can still be added and it might be good to test these with otk directly before merging, but they're mostly functionally complete.

### CODE REVIEW

File: internal/otkdisk/partition.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+type Const struct {
+	KernelOptsList []string `json:"kernel_opts_list"`
+
+	// PartitionMap is generated for convenient indexing of certain partitions
+	// with predictable names in otk, such as
+	// "filesystem.partition_map.boot.uuid"
+	PartitionMap map[string]Partition `json:"partition_map"`
+
+	// Internal representation of the full partition table.
```
Comment (mvo5) [inline internal/otkdisk/partition.go:22]: Maybe we could add  here
```
	// Internal representation of the full partition table. The representation is internal to the partition tools and should not be used by otk directly. No external API gurantees about the conent
```
or something like this?

File: pkg/disk/partition.go [hunk 2]
```diff
@@ -18,6 +19,44 @@ type Partition struct {
 	Payload Entity
 }
 
+// XXX: this needs refactoring and tests, it's kinda complicated
```
Comment (mvo5) [inline pkg/disk/partition.go:22]: I will work on this in a followup/parallel PR, this is tested indirectly via the otk-make-* tests but of course a direct test is much better and will get added. Note that this is only used by the otk-make-* tooling right now so the risk to merge in the current state is low.

**Review by mvo5 — APPROVED** [review]
Looks good to me (but my vote counts only +0.5 as I wrote parts of this)

File: internal/otkdisk/partition.go (hunk 1 above)
Comment (achilleas-k) [reply internal/otkdisk/partition.go:22]: Updated the docstring.

File: pkg/disk/partition.go (hunk 2 above)
Comment (mvo5) [reply pkg/disk/partition.go:22]: I started this in https://github.com/osbuild/images/pull/737

**Review by mvo5 — DISMISSED** [review]
Still looks good to me (but only 0.5 because I wrote some of the code)

File: cmd/otk-gen-partition-table/main.go [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+	"github.com/osbuild/images/internal/buildconfig"
+	"github.com/osbuild/images/internal/cmdutil"
+	"github.com/osbuild/images/internal/common"
+	"github.com/osbuild/images/internal/otkdisk"
+	"github.com/osbuild/images/pkg/blueprint"
+	"github.com/osbuild/images/pkg/disk"
+	"github.com/osbuild/images/pkg/osbuild"
+)
+
+type Input struct {
```
Comment (ondrejbudai) [inline cmd/otk-gen-partition-table/main.go:19]: Since this basically becomes our public API, I think that all of these have to be documented with docstrings.

File: cmd/otk-gen-partition-table/main.go [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+	Properties InputProperties   `json:"properties"`
+	Partitions []*InputPartition `json:"partitions"`
+
+	Modifications InputModifications `json:"modifications"`
+}
+
+type InputProperties struct {
+	UEFI        InputUEFI `json:"uefi"`
+	BIOS        bool      `json:"bios"`
+	Type        string    `json:"type"`
```
Comment (ondrejbudai) [inline cmd/otk-gen-partition-table/main.go:29]: Can we make this an enum?

Comment (achilleas-k) [reply cmd/otk-gen-partition-table/main.go:29]: I considered it but since this is an input data structure I think making it an enum string doesn't give us much benefit.  But we should validate the values somehow so perhaps an enum makes that more explicit.

**Review by ondrejbudai — CHANGES_REQUESTED** [review]
Awesome! I played with this a bit, and found several tiny issues, see inline. :)

File: pkg/disk/partition.go [hunk 5]
```diff
@@ -18,6 +19,44 @@ type Partition struct {
...
+func (p *Partition) UnmarshalJSON(data []byte) error {
+	type partAlias Partition
+	var aux struct {
+		partAlias
+		Payload struct {
```
[the remark this answers is not in the document]
Comment (mvo5) [reply pkg/disk/partition.go:27]: Thank you! This part of the code will be replaced with https://github.com/osbuild/images/pull/737 and I added code to handle this case now (it was not doing that before, so a huge *thank you* for providing this example).

File: cmd/otk-gen-partition-table/main.go [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+	DefaultSize string    `json:"default_size"`
+	UUID        string    `json:"uuid"`
+
+	SectorSize uint64 `json:"sector_size"`
+}
+
+type InputUEFI struct {
+	Size string `json:"size"`
```
[the remark this answers is not in the document]
Comment (mvo5) [reply cmd/otk-gen-partition-table/main.go:37]: I was thinking about this and the `BIOSBootPartition` and wonder if we should change the structure a little bit as both of those are about automatically creating partitions, maybe something like:
```go
diff --git a/cmd/otk-gen-partition-table/main.go b/cmd/otk-gen-partition-table/main.go
index 496914894..6d78fb2f9 100644
--- a/cmd/otk-gen-partition-table/main.go
+++ b/cmd/otk-gen-partition-table/main.go
@@ -27,17 +27,20 @@ type Input struct {
 
 // InputProperties contains global properties of the partition table
 type InputProperties struct {
-       UEFI        InputUEFI        `json:"uefi"`
-       BIOS        bool             `json:"bios"`
        Type        otkdisk.PartType `json:"type"`
        DefaultSize string           `json:"default_size"`
        UUID        string           `json:"uuid"`
+       Create      InputCreate      `json:"create"`
 
        SectorSize uint64 `json:"sector_size"`
 }
 
-// InputUEFI contains the uefi specific bits of the partition input
-type InputUEFI struct {
+// InputAutogenerate contains details what partitions to auto-generate
+type InputCreate struct {
+       BIOSBootPartition bool   `json:"bios-boot-partition"`
+       EspPartition      bool   `json:"esp-partition"`
+       EspPartitionSize  string `json:"esp-partition-size"`
+
        Size string `json:"size"`
 }
 
@@ -111,7 +114,7 @@ func makePartitionTableFromOtkInput(input *Input) (*disk.PartitionTable, error)
                Type:       string(input.Properties.Type),
                SectorSize: input.Properties.SectorSize,
        }
-       if input.Properties.BIOS {
+       if input.Properties.Create.BIOSBootPartition {
                if len(pt.Partitions) > 0 {
                        return nil, fmt.Errorf("internal error: bios partition *must* go first")
                }
@@ -122,8 +125,13 @@ func makePartitionTableFromOtkInput(input *Input) (*disk.PartitionTable, error)
                        UUID:     disk.BIOSBootPartitionUUID,
                })
        }
-       if input.Properties.UEFI.Size != "" {
-               uintSize, err := common.DataSizeToUint64(input.Properties.UEFI.Size)
+       if input.Properties.Create.EspPartition {
+               size := input.Properties.Create.EspPartitionSize
+               if size == "" {
+                       // XXX: is this a good default?
+                       size = "1 Gb"
+               }
+               uintSize, err := common.DataSizeToUint64(size)
                if err != nil {
                        return nil, err
                }

```
alternative names might be `InputAutoCreate`, `InputAutoGenerate` or similar.

File: cmd/otk-gen-partition-table/main.go [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+type Input struct {
+	Properties InputProperties   `json:"properties"`
+	Partitions []*InputPartition `json:"partitions"`
+
+	Modifications InputModifications `json:"modifications"`
+}
+
+type InputProperties struct {
+	UEFI        InputUEFI `json:"uefi"`
+	BIOS        bool      `json:"bios"`
```
[the remark this answers is not in the document]
Comment (mvo5) [reply cmd/otk-gen-partition-table/main.go:28]: See above.

File: cmd/otk-gen-partition-table/main.go [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+			})
+		}
+	}
+
+	for _, part := range input.Partitions {
+		uintSize, err := common.DataSizeToUint64(part.Size)
+		if err != nil {
+			return nil, err
+		}
+		pt.Partitions = append(pt.Partitions, disk.Partition{
```
[the remark this answers is not in the document]
Comment (mvo5) [reply cmd/otk-gen-partition-table/main.go:154]: Indeed, good point!

File: cmd/otk-make-partition-stages/main.go [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+type Input = otkdisk.Data
+
+func makeImagePrepareStages(inp Input, filename string) (stages []*osbuild.Stage, err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			err = fmt.Errorf("cannot generate image prepare stages: %v", r)
+		}
+	}()
+
+	// rhel7 uses PTSgdisk, if we ever need to support this, make this
```
[the remark this answers is not in the document]
Comment (mvo5) [reply cmd/otk-make-partition-stages/main.go:23]: AFAICT we are safe here, if we need to support another tool we would do something like:
```go
diff --git a/cmd/otk-make-partition-stages/main.go b/cmd/otk-make-partition-stages/main.go
index 66c25349a..6d6f76bff 100644
--- a/cmd/otk-make-partition-stages/main.go
+++ b/cmd/otk-make-partition-stages/main.go
@@ -20,10 +20,7 @@ func makeImagePrepareStages(inp Input, filename string) (stages []*osbuild.Stage
                }
        }()
 
-       // rhel7 uses PTSgdisk, if we ever need to support this, make this
-       // configurable
-       partTool := osbuild.PTSfdisk
-       stages = osbuild.GenImagePrepareStages(inp.Const.Internal.PartitionTable, inp.Const.Filename, partTool)
+       stages = osbuild.GenImagePrepareStages(inp.Const.Internal.PartitionTable, inp.Const.Filename, osbuild.PartTool(inp.Const.Internal.PartTool))
        return stages, nil
 }
 
diff --git a/internal/otkdisk/partition.go b/internal/otkdisk/partition.go
index 37b649c5c..33c199885 100644
--- a/internal/otkdisk/partition.go
+++ b/internal/otkdisk/partition.go
@@ -50,6 +50,7 @@ type Partition struct {
 // "otk-{gen,make}-*" tools for their data exchange.
 type Internal struct {
        PartitionTable *disk.PartitionTable `json:"partition-table"`
+       PartTool       string               `json:"part-tool"`
 }
 
 // PartType represents a partition type
```
and from otk-gen-* we would set it (or move it out of internal if it should be something the user should control). Or am I misunderstanding the comment here?

File: pkg/disk/partition.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (mvo5) [reply pkg/disk/partition.go:27]: Might be easiest to review #737 first and then I can just rebase on top of this and drop this initial and slightly naive implementation for the unmarshal.

**Review by mvo5 — APPROVED** [review]
I rebased and removed the old marshal code (now that #775 is in) - this is ready now. Please consider my approval only a +0.5 as I wrote good chunks of this.

**Review by achilleas-k — COMMENTED** [review]
I reviewed the most recent changes and it all LGTM.  I'd approve but I can't since I'm the author/owner here.

The weak rng usage in tests needs comments for linter skipping.

I think we should go ahead and merge this without requiring further detailed reviews.  The PR isn't in any critical path yet and I think bugs, edge cases, and usability issues will surface once we start using the features in otk.  We can iterate on the implementation then.

File: cmd/otk-gen-partition-table/main.go [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+			return 0, err
+		}
+	}
+	if input.Modifications.MinDiskSize != "" {
+		modMinSize, err = common.DataSizeToUint64(input.Modifications.MinDiskSize)
+		if err != nil {
+			return 0, err
+		}
+	}
+	// TODO: use max() once we move to go1.21
```
Comment (ondrejbudai) [inline cmd/otk-gen-partition-table/main.go:193]: The future is now.

File: cmd/otk-make-partition-mounts-devices/main.go [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+type Output struct {
+	RootMountName string                    `json:"root_mount_name"`
+	Mounts        []osbuild.Mount           `json:"mounts"`
+	Devices       map[string]osbuild.Device `json:"devices"`
+}
+
+func run(r io.Reader, w io.Writer) error {
+	var inp Input
+	if err := json.NewDecoder(r).Decode(&inp); err != nil {
+		return err
```
Comment (ondrejbudai) [inline cmd/otk-make-partition-mounts-devices/main.go:24]: Wrap me. :)

File: cmd/otk-make-partition-stages/main.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (ondrejbudai) [reply cmd/otk-make-partition-stages/main.go:23]: But is an omnifest author supposed to touch the internal fields? Feels fishy... That's why I proposed adding another layer around `internal`. Alternatively, we can put it into `InputProperties` and pass it from `gen-partition-table` to `make-partition-stages` in `internal`. Might be simpler for the user.

File: internal/otkdisk/partition.go [hunk 12]
```diff
@@ -0,0 +1,70 @@
+package otkdisk
+
+import (
+	"fmt"
+
+	"golang.org/x/exp/slices"
```
Comment (ondrejbudai) [inline internal/otkdisk/partition.go:6]: The future is indeed now, so we can just use `slices` here. Might be worth doing this in a follow-up and changing it globally. :)

**Review by mvo5 — DISMISSED** [review]
Fwiw, I agree

**Review by ondrejbudai — COMMENTED** [review]
I'm loving this! :) I have some concerns about field names, missing documentation and hard-to-understand error messages. However, functionality looks alright to me. :) I don't think that anything is a real blocker here, but it might be a good idea to adjust the field names before merging, so we can rely on them from now on.

Also, we will need an fstab stage generator. Once again, not sure if this should be a part of this PR, or a follow-up.

File: cmd/otk-gen-partition-table/main.go [hunk 13]
```diff
@@ -0,0 +1,268 @@
...
+	if input.Properties.Create.EspPartition {
+		size := input.Properties.Create.EspPartitionSize
+		if size == "" {
+			size = "1 GiB"
+		}
+		uintSize, err := common.DataSizeToUint64(size)
+		if err != nil {
+			return nil, err
+		}
+		if uintSize > 0 {
```
[the remark this answers is not in the document]
Comment (achilleas-k) [reply cmd/otk-gen-partition-table/main.go:136]: Do we want to error or treat 0 as a "don't create"?

**Review by mvo5 — DISMISSED** [review]
Thanks for the extra tweaks/fixes!

**Review by ondrejbudai — APPROVED** [review]
Thanks!

**Review by supakeen — DISMISSED** [review]

**Comment by supakeen** [comment]
I've resolved all conversations in this PR so it can go into the merge since it has multiple acks since those conversations were created (a lot of them *should* go into followups!).

**Review by mvo5 — DISMISSED** [review]
I had to rebase as there was a rename in the code that made this branch stop compiling (genMountsDevicesFromPt -> GenMountsDevicesFromPT)

**Review by mvo5 — APPROVED** [review]
